compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
I don't think the AM handler will ever need to do anything else to satisfy the CAS loop, and I would guess that  endCounts updates will be relatively uncontested so we shouldn't spin much. I think it's a similar case as with gemini. The only people who can update the endcount are local tasks, or remote tasks through the AM handler